### ISSUE TITLE
What instances should Haddock show?

### ISSUE THREAD

**Issue by harpocrates** [haskell/haddock#916]
We've historically had problems with Haddock not showing enough instances. Our current solution is to display what GHCi would if you loaded all of the modules in the package being processed. Unfortunately, once the Hi Haddock changes land, I'm not sure we're going to be able to replicate that logic.

Considering this as an opportunity, the question of the day is: **what modules do we want to load before collecting instances?**. If we load too many modules, we risk ending up with a bloated list of instances, but if we load too few modules we might omit instances a user would expect to see.

---

### Loading no extra modules

The Hi-Haddock branch of today loads no extra modules before looking up instances. That means it ends up missing some instances. Here is what the instance list for `Array` from the `array` package looks like ([link to is what today's Haddock produces](https://hackage.haskell.org/package/array-0.5.2.0/docs/Data-Array.html#t:Array)):

<img width="558" alt="screen shot 2018-08-22 at 10 47 32 pm" src="https://user-images.githubusercontent.com/10766081/44506926-703cc400-a65d-11e8-9472-11af57e4b7c1.png">

Here is what memory usage (produced with `haddock +RTS -hc -RTS ...`) looks like for `ghc`-the-library.

<img width="693" alt="screen shot 2018-08-22 at 10 56 56 pm" src="https://user-images.githubusercontent.com/10766081/44507262-b6deee00-a65e-11e8-8950-293115d9a48e.png">

### Loading all transitively-depended-upon modules

For the sake of motivating discussion, I've hacked on the Hi-Haddock branch to make it load every module transitively depended on (esp. from external packages). See how it pulls in a lot more instances for `Array` from the `array` package:

<img width="594" alt="screen shot 2018-08-22 at 10 47 40 pm" src="https://user-images.githubusercontent.com/10766081/44507305-e0981500-a65e-11e8-8d70-b3fe69ad3888.png">

Memory usage when Haddock-ing `ghc`-the-library.

<img width="698" alt="screen shot 2018-08-22 at 10 59 12 pm" src="https://user-images.githubusercontent.com/10766081/44507398-4be1e700-a65f-11e8-8b4d-8229a381c07e.png">

**Comment by harpocrates** [comment]
One potentially reasonable option I've just thought of: we could load only modules which are _visible_ (either directly, or through re-exports). That way, we'll pick up all of the instances from direct dependencies, without also collecting probably un-hyperlinkable instances from transitive dependencies. Plus that means that packages with gigantic transitive dependencies but small direct dependencies won't have to load a ton of likely uninteresting modules.

I tried this out at https://github.com/harpocrates/haddock/commit/173f079d517a9e6f1a0b54141d71a3842c5ee383 and the results look about right. Here's `Array` again (notice that we've picked up the `Data` instance, but not the indirect `NFData`/`NFData1`/`NFData2` and `Binary` instances):

<img width="557" alt="screen shot 2018-08-23 at 1 08 55 am" src="https://user-images.githubusercontent.com/10766081/44513034-28279c80-a671-11e8-9fc4-99648b1e361f.png">

**Comment by sjakobi** [comment]
That's a great question!

Yet another variant would be to list those instances that are available when the module is imported:

```
λ import Data.Array
λ :i Array
type role Array nominal representational
data Array i e
  = GHC.Arr.Array !i !i {-# UNPACK #-}Int (GHC.Prim.Array# e)
  	-- Defined in ‘GHC.Arr’
instance (Ix i, Eq e) => Eq (Array i e) -- Defined in ‘GHC.Arr’
instance Functor (Array i) -- Defined in ‘GHC.Arr’
instance (Ix i, Ord e) => Ord (Array i e) -- Defined in ‘GHC.Arr’
instance (Ix a, Show a, Show b) => Show (Array a b)
  -- Defined in ‘GHC.Arr’
instance (Ix a, Read a, Read b) => Read (Array a b)
  -- Defined in ‘GHC.Read’
instance Foldable (Array i) -- Defined in ‘Data.Foldable’
instance Ix i => Traversable (Array i)
  -- Defined in ‘Data.Traversable’
```

Notably that misses the `IArray` and `Data` instances.

From the options you listed, I think I'd personally prefer the "visible modules" option.

**Comment by harpocrates** [comment]
I've opened https://phabricator.haskell.org/D5100 related to this. If that gets merged, we should be able to use https://github.com/harpocrates/haddock/commit/173f079d517a9e6f1a0b54141d71a3842c5ee383 almost as is.

**Comment by harpocrates** [comment]
> I've opened https://phabricator.haskell.org/D5100 related to this. If that gets merged, we should be able to use harpocrates@173f079 almost as is.

The phab diff got merged.

I've also given this more thought and I think there is one more set of modules we will want to ensure are loaded: the transitive closure of orphan modules / family instance modules (this will ensure we catch orphan instances from some transitively depended upon package). I'm not sure these aren't already loaded by the time we get to looking up instances. If they aren't we could manually track these from `dep_orphs . mi_deps` of each interface we process.

**Comment by sjakobi** [comment]
I think we might now show instances that we shouldn't:

![image](https://user-images.githubusercontent.com/3664523/47380652-38560980-d6fe-11e8-9834-e018099423e3.png)

This is from the `Prelude` docs. The problem IMHO is that while the `Show` class and many of its instances are in scope here, that particular `Show` instance isn't. I think this would be confusing to many users, especially newcomers.

**Comment by harpocrates** [comment]
Perhaps we should have a way of annotating instances with info about whether or not they are visible _from module whose docs you are looking at_.

That said, I think we should show this instance in some way (I didn't even realize it existed and I don't know how to find it in today's Haddock!), and this is a pretty good place for it. This particular instance should probably also have some Haddock comment on it. Maybe we could move the module's description onto the instance?

**Comment by harpocrates** [comment]
@sjakobi I think everything in this issue has been fixed in the Hi Haddock branch, right? https://github.com/haskell/haddock/issues/916#issuecomment-432360232 seems like the only thing that is left, and that probably belongs as a separate feature request. Is that correct?

**Comment by harpocrates** [comment]
Fixed in d10fb1292157cb6f5e69cec88650d314d09bd95d.
